### PR TITLE
[JENKINS-73164] Fix record issues step ignores source code retention parameter

### PR DESCRIPTION
**Pull request by sirine707**
- Fix the `recordIssues` step to correctly honor the `sourceCodeRetention` parameter.
- Added a test to verify that source code is not copied when sourceCodeRetention: 'NEVER'` is set.

### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java [hunk 1]
```diff
@@ -99,4 +100,28 @@ void shouldRecordOutputOfParallelSteps() {
         assertThat(first.getResult().getIssues()).hasSize(5);
         assertThat(second.getResult().getIssues()).hasSize(3);
     }
+
+    /**
+     * Verifies that source files are not retained in the Jenkins build folder when
+     * the 'sourceCodeRetention' policy is set to 'NEVER'
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:106]:
```suggestion
     * the 'sourceCodeRetention' policy is set to 'NEVER'.
```

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java [hunk 2]
```diff
@@ -99,4 +100,28 @@ void shouldRecordOutputOfParallelSteps() {
...
+
+        WorkflowJob project = createPipeline();
+
+        createFileInAgentWorkspace(agent, project, "Test.java", JAVA_CONTENT);
+
+        project.setDefinition(new CpsFlowDefinition("node('agent') {\n"
+                + "    echo '[javac] Test.java:39: warning: Test Warning'\n"
+                + "    recordIssues tool: java(), sourceCodeRetention: 'NEVER'\n"
+                + "}", true));
+
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:120]: Can you check here that the console log reports that no files are stored? Similar to 

```
assertThat(getConsoleLog(result)).contains("1 copied", "0 not in workspace", "0 not-found", "0 with I/O error");
```

there should be a log message that shows that source code files are not stored...

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java [hunk 3]
```diff
@@ -99,4 +100,28 @@ void shouldRecordOutputOfParallelSteps() {
...
+        AnalysisResult result = scheduleSuccessfulBuild(project);
+        assertThat(result).hasNoErrorMessages();
+        assertThat(result).hasTotalSize(1);
+        assertThat(getSourceCode(result, 0)).contains("FileNotFoundException");
+    }
+
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:126]:
```suggestion
```

**Review by uhafner — CHANGES_REQUESTED** [review]
Almost done. Thanks for the PR!

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java (hunk 2 above)
Comment (sirine707) [reply plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:120]: thank you @KalleOlaviNiemitalo for the insights, @uhafner what do you think about this ?

Comment (uhafner) [reply plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:120]: I think it would be better to add a logging statement in the else part so everybody can see that the parameter is correctly set. Something like "Skipping copying of affected files"

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java [hunk 4]
```diff
@@ -87,11 +77,11 @@ enum PostProcessingMode {
...
-            final FilePath workspace, final Set<String> sourceDirectories,
-            final SourceCodeRetention sourceCodeRetention, final Run<?, ?> run,
-            final FilePath jenkinsRootDir, final TaskListener listener,
-            final String scm, final BlameMode blameMode, final PostProcessingMode postProcessingMode,
-            final boolean quiet) {
+                  final FilePath workspace, final Set<String> sourceDirectories,
+                  final SourceCodeRetention sourceCodeRetention, final Run<?, ?> run,
+                  final FilePath jenkinsRootDir, final TaskListener listener,
+                  final String scm, final BlameMode blameMode, final PostProcessingMode postProcessingMode,
+                  final boolean quiet) {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java:84]: Please avoid unrelated changes

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java [hunk 5]
```diff
@@ -139,8 +129,7 @@ private AnnotatedReport postProcessReport(final Report report) throws IOExceptio
             AnnotatedReport result = workspace.act(createPostProcessor(report));
             copyAffectedFiles(result.getReport(), createAffectedFilesFolder(result.getReport()));
             return result;
-        }
-        else {
+        } else {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java:132]: Please avoid unrelated changes

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java [hunk 6]
```diff
@@ -164,8 +153,7 @@ private Blamer createBlamer(final Report report) {
         if (blameMode == BlameMode.DISABLED) {
             report.logInfo("Skipping SCM blames as requested");
             return new NullBlamer();
-        }
-        else {
+        } else {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java:156]: Please avoid unrelated changes

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java [hunk 7]
```diff
@@ -189,6 +177,8 @@ private void copyAffectedFiles(final Report report, final FilePath buildFolder)
             permittedSourceDirectories.add(workspace.getRemote());
             new AffectedFilesResolver().copyAffectedFilesToBuildFolder(
                     report, workspace, permittedSourceDirectories, buildFolder);
+        } else {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java:180]: Check your codingstyle settings.

```suggestion
        } 
        else {
```

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java [hunk 8]
```diff
@@ -346,8 +331,7 @@ private void resolvePackageNames(final Report report) {
             try {
                 PackageNameResolver resolver = new PackageNameResolver();
                 resolver.run(report, getCharset());
-            }
-            catch (InvalidPathException exception) {
+            } catch (InvalidPathException exception) {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java:334]: It seems you reformatted the whole source with the wrong style

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java [hunk 9]
```diff
@@ -90,13 +89,36 @@ void shouldRecordOutputOfParallelSteps() {
         if (JAVA_ID.equals(actions.get(0).getId())) {
             first = actions.get(0);
             second = actions.get(1);
-        }
-        else {
+        } else {
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:92]: see above

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java [hunk 10]
```diff
@@ -90,13 +89,36 @@ void shouldRecordOutputOfParallelSteps() {
...
+
+        project.setDefinition(new CpsFlowDefinition("node('agent') {\n"
+                + "    echo '[javac] Test.java:39: warning: Test Warning'\n"
+                + "    recordIssues tool: java(), sourceCodeRetention: 'NEVER'\n"
+                + "}", true));
+
+        AnalysisResult result = scheduleSuccessfulBuild(project);
+        assertThat(result).hasNoErrorMessages();
+        assertThat(result).hasTotalSize(1);
+        assertThat(getConsoleLog(result)).contains("Skipping copying of affected files");
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:121]: That looks good now!

**Review by uhafner — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java (hunk 8 above)
Comment (sirine707) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java:334]: its the intellij on gitpod who keeps doing it , I disabled the default settings but each time I have the same update of white spaces when I added the else block,  
do I update the formatting of else and catch to its original and I commit again ?

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java (hunk 10 above)
Comment (sirine707) [reply plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsOnAgentITest.java:121]: Thank you for the update !

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java (hunk 8 above)
Comment (uhafner) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesScanner.java:335]: Compare your changes with main and revert everything that is unrelated.

**Review by uhafner — APPROVED** [review]

**Comment by uhafner** [comment]
Thanks!
